packages/teleport/src/components/StepSlider/StepSlider.tsx

```diff
@@ -0,0 +1,258 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import React, { useState, useRef, useEffect, RefObject } from 'react';
+import { TransitionGroup, CSSTransition } from 'react-transition-group';
+import styled from 'styled-components';
+
+export default function StepSlider<T>(props: Props<T>) {
+  const {
+    flows,
+    currFlow,
+    onSwitchFlow,
+    render,
+    tDuration = 500,
+    ...rest
+  } = props;
+
+  // step defines the current step we are in the current flow.
+  const [step, setStep] = useState(0);
+  const [height, setHeight] = useState(0);
+  const [className, setClassName] = useState<'next' | 'prev' | ''>('');
+
+  // preMount is used to invisibly render the next view so we
+  // can get its height. This height is needed in advance
+  // so the height can be animated along with the transform transitions.
+  const [preMount, setPreMount] = useState(false);
+
+  // rootRef is used to set the height on initial render.
+  // Needed to animate the height on initial transition.
+  const rootRef = useRef<HTMLElement>();
+
+  // preMountState is used to hold the latest pre mount data.
+  // useState's could not be used b/c they became stale for
+  // our func 'setHeightOnPreMount'.
+  const preMountState = useRef<{ step: number; flow: keyof T }>({} as any);
+
+  // Sets the initial height.
+  useEffect(() => {
+    if (!rootRef?.current) return;
+    const { height } = rootRef.current.getBoundingClientRect();
+    setHeight(height);
+  }, []);
+
+  // After pre mount, we can calculate the exact height of the next step.
+  // After calculating height, we increment the step to trigger the
+  // animations.
+  const setHeightOnPreMount = (node: HTMLElement) => {
+    if (node !== null) {
+      setHeight(node.getBoundingClientRect().height);
+      setStep(preMountState.current.step);
+      setPreMount(false);
+
+      if (preMountState.current.flow) {
+        onSwitchFlow(preMountState.current.flow);
+      }
+    }
+  };
+
+  function getStep(View: StepComponent, requirePreMount = false) {
+    return (
+      <View
+        key={step}
+        refCallback={requirePreMount ? setHeightOnPreMount : null}
+        onNext={() => {
+          preMountState.current.step = step + 1;
+          setPreMount(true);
+          setClassName('next');
+        }}
+        onPrev={() => {
+          preMountState.current.step = step - 1;
+          setPreMount(true);
+          setClassName('prev');
+        }}
+        onSwitchFlow={(flow, isBack = false) => {
```

Comment:
`isBack` feels like a bit of a bandaid

If switching flows always resets that flows step to 0, why would the animation be anything other than `next`?

Reply:
it was to provide a different animation if user wants to switch back to another flow:
e.g. if we had:
  - primary flow
  - secondary flow

User will start at primary:
- user clicks secondary flow (applies `prev` animation, on second thought this should be `next` animation instead)
- user is now on secondary flow and wants to go "back" to the primary flow, so `isBack` is set to true, and we apply `prev` animation